src/popen.c

```diff
@@ -113,6 +113,42 @@ FILE *mypopen(const char *command, volatile pid_t *pidptr)
     exit(1);
 }
 
+FILE *mypopene(const char *command, volatile pid_t *pidptr, char **env) {
+    int pipefd[2];
+
+    if(pipe(pipefd) == -1)
+        return NULL;
+
+    int pid = fork();
+    if(pid == -1) {
+        close(pipefd[PIPE_READ]);
+        close(pipefd[PIPE_WRITE]);
+        return NULL;
+    }
+    if(pid != 0) {
+        // the parent
+        *pidptr = pid;
+        close(pipefd[PIPE_WRITE]);
+        FILE *fp = fdopen(pipefd[PIPE_READ], "r");
+        return(fp);
+    }
+    // the child
+
+    // close all files
+    int i;
+    for(i = (int) (sysconf(_SC_OPEN_MAX) - 1); i > 0; i--)
+        if(i != STDIN_FILENO && i != STDERR_FILENO && i != pipefd[PIPE_WRITE]) close(i);
+
+    // move the pipe to stdout
+    if(pipefd[PIPE_WRITE] != STDOUT_FILENO) {
+        dup2(pipefd[PIPE_WRITE], STDOUT_FILENO);
+        close(pipefd[PIPE_WRITE]);
+    }
+
+    execle("/bin/sh", "sh", "-c", command, NULL, env);
```

Comment:
The risk of command injection could be reduced by eliminating the "/bin/sh" here: It's a whole extra shell process that's not doing any useful work except splitting the `command` apart into the pieces from which we've just assembled it. If the helper script path and its arguments were passed in as separate strings (or an array of them, for `execvple`), then no shell ever has to parse them (assuming all uses of them in the helper script itself are properly quoted).